### PR TITLE
8292675: Add identity transformation for removing redundant AndV/OrV nodes

### PR DESCRIPTION
**Pull request by Bhavana-Kilambi**
Recently we found that the rotate left/right benchmarks with vectorapi
emit a redundant "and" instruction on both aarch64 and x86_64 machines
which can be done away with.  For example - and(and(a, b), b) generates
two "and" instructions which can be reduced to a single "and" operation-
and(a, b) since "and" (and "or") operations are commutative and
idempotent in nature.  This can help improve performance for all those
workloads which have multiple "and"/"or" operations with the same value
by reducing them to fewer "and"/"or" operations accordingly.

This patch adds the following transformations for vector logical
operations - AndV and OrV :

```
(OpV (OpV a b) b) => (OpV a b)
(OpV (OpV a b) a) => (OpV a b)
(OpV (OpV a b m1) b m1) => (OpV a b m1)
(OpV (OpV a b m1) a m1) => (OpV a b m1)
(OpV a (OpV a b)) => (OpV a b)
(OpV b (OpV a b)) => (OpV a b)
(OpV a (OpV a b m) m) => (OpV a b m)
```
where Op = "And", "Or"

Links for benchmarks tested are given below :-
https://github.com/openjdk/panama-vector/blob/2aade73adeabdf6a924136b17fd96ccc95c1d160/test/micro/org/openjdk/bench/jdk/incubator/vector/operation/IntMaxVector.java#L728
https://github.com/openjdk/panama-vector/blob/2aade73adeabdf6a924136b17fd96ccc95c1d160/test/micro/org/openjdk/bench/jdk/incubator/vector/operation/IntMaxVector.java#L764
https://github.com/openjdk/panama-vector/blob/2aade73adeabdf6a924136b17fd96ccc95c1d160/test/micro/org/openjdk/bench/jdk/incubator/vector/operation/LongMaxVector.java#L728
https://github.com/openjdk/panama-vector/blob/2aade73adeabdf6a924136b17fd96ccc95c1d160/test/micro/org/openjdk/bench/jdk/incubator/vector/operation/LongMaxVector.java#L764

Before this patch, the disassembly for one these testcases
(IntMaxVector.ROR) for Neon is shown below :
 ```
  ldr     q16, [x12, #16]
  and     v16.16b, v16.16b, v20.16b
  and     v16.16b, v16.16b, v20.16b
  add     x12, x16, x11
  sub     v17.4s, v21.4s, v16.4s
  ...
  ...
```

After this patch, the disassembly for the same testcase above is shown
below :
```
  ldr     q16, [x12, #16]
  and     v16.16b, v16.16b, v20.16b
  add     x12, x16, x11
  sub     v17.4s, v21.4s, v16.4s
  ...
  ...
```

The other tests also emit an extra "and" instruction as shown above for
the vector ROR/ROL operations.

Below are the performance results for the vectorapi rotate tests (tests
given in the links above) with this patch on aarch64 and x86_64 machines
(for int and long types) -

```
Benchmark                aarch64   x86_64
IntMaxVector.ROL         25.57%    26.09%
IntMaxVector.ROR         23.75%    24.15%
LongMaxVector.ROL        28.91%    28.51%
LongMaxVector.ROR        16.51%    29.11%

```

The percentage indicates the percent gain/improvement in performance
(ops/ms) with this patch over the master build without this patch. The
machine descriptions are given below -
aarch64 - 128-bit aarch64 machine
x86_64  - 256-bit x86 machine

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292675](https://bugs.openjdk.org/browse/JDK-8292675): Add identity transformation for removing redundant AndV/OrV nodes

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Nick Gasson](https://openjdk.org/census#ngasson) (@nick-arm - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10163/head:pull/10163` \
`$ git checkout pull/10163`

Update a local copy of the PR: \
`$ git checkout pull/10163` \
`$ git pull https://git.openjdk.org/jdk pull/10163/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10163`

View PR using the GUI difftool: \
`$ git pr show -t 10163`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10163.diff">https://git.openjdk.org/jdk/pull/10163.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/vectornode.cpp [hunk 1]
```diff
@@ -1879,6 +1879,39 @@ Node* ReverseVNode::Identity(PhaseGVN* phase) {
...
+static Node* redundant_logical_identity(Node* n) {
+  Node* n1 = n->in(1);
+  // (OperationV (OperationV src1 src2) src1) => (OperationV src1 src2)
+  // (OperationV (OperationV src1 src2) src2) => (OperationV src1 src2)
+  // (OperationV (OperationV src1 src2 m1) src1 m1) => (OperationV src1 src2 m1)
+  // (OperationV (OperationV src1 src2 m1) src2 m1) => (OperationV src1 src2 m1)
+  if (n->Opcode() == n1->Opcode()) {
+    if (((!n->is_predicated_vector() && !n1->is_predicated_vector()) ||
+       (n->is_predicated_vector() && n1->is_predicated_vector() &&
+       n->in(3) == n1->in(3))) && (n->in(2) == n1->in(1) || n->in(2) == n1->in(2))) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/vectornode.cpp:1893]: This condition is hard to parse, I would suggest:

```suggestion
    if (((!n->is_predicated_vector() && !n1->is_predicated_vector()) ||
         ( n->is_predicated_vector() &&  n1->is_predicated_vector() && n->in(3) == n1->in(3))) &&
        (n->in(2) == n1->in(1) || n->in(2) == n1->in(2))) {
```

File: src/hotspot/share/opto/vectornode.cpp [hunk 2]
```diff
@@ -1879,6 +1879,39 @@ Node* ReverseVNode::Identity(PhaseGVN* phase) {
...
+      return n1;
+    }
+  }
+
+  Node* n2 = n->in(2);
+  if (n->Opcode() == n2->Opcode()) {
+    // (OperationV src1 (OperationV src1 src2)) => OperationV(src1, src2)
+    // (OperationV src2 (OperationV src1 src2)) => OperationV(src1, src2)
+    if (!n->is_predicated_vector() && !n2->is_predicated_vector() &&
+       (n->in(1) == n2->in(1) || n->in(1) == n2->in(2))) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/vectornode.cpp:1903]:
```suggestion
    if (!n->is_predicated_vector() && !n2->is_predicated_vector() &&
        (n->in(1) == n2->in(1) || n->in(1) == n2->in(2))) {
```

File: src/hotspot/share/opto/vectornode.cpp [hunk 3]
```diff
@@ -1879,6 +1879,39 @@ Node* ReverseVNode::Identity(PhaseGVN* phase) {
...
+      return n2;
+    // (OperationV src1 (OperationV src1 src2 m1) m1) => OperationV(src1 src2 m1)
+    } else if (n->is_predicated_vector() && n2->is_predicated_vector() &&
+               n->in(3) == n2->in(3) && n->in(1) == n2->in(1)) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/vectornode.cpp:1907]: I think this should be merged into line 1902. Why did you omit the `(OperationV src2 (OperationV src1 src2 m1) m1) => OperationV(src1 src2 m1)` case?

**Review by TobiHartmann — CHANGES_REQUESTED** [review]

File: src/hotspot/share/opto/vectornode.cpp (hunk 3 above)
Comment (Bhavana-Kilambi) [reply src/hotspot/share/opto/vectornode.cpp:1907]: Hi, thank you for your review comments. I will make the suggested changes and upload a new patch.
Regarding omitting this condition - `(OperationV src2 (OperationV src1 src2 m1) m1) => This would result in a different result as compared to `(OperationV src1 src2 m1)`. So the inner `OperationV` would result in `src1` being copied for unmasked lanes and the outer `OperationV` would result in `src2` being copied to the final result for unmasked lanes and thus the results for both the operations is different. So for masked lanes, the result is Operation(src1, src2) but not so for the unmasked lanes. So we need to have the first arguments of both `OperationV` to be same so that the unmasked lanes would then have the same value in the end and that can then be optimized to a single `OperationV` node.

Comment (TobiHartmann) [reply src/hotspot/share/opto/vectornode.cpp:1907]: Right, I missed that. Would probably not hurt adding a corresponding comment. Thanks!

Comment (Bhavana-Kilambi) [reply src/hotspot/share/opto/vectornode.cpp]: Thanks. I have added a comment for this case.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. Testing in our CI passed.

**Review by nick-arm — APPROVED** [review]
